### PR TITLE
ADBDEV-4954 gpexpand: skip expanded/broken tables when redistributing

### PR DESCRIPTION
**Pull request by dnskvlnk**
This commit is a cherry-pick of https://github.com/arenadata/gpdb/commit/6b7e4c3dbb586fde520d31e19ff4a781e8dcfdaa.
In summary `gpexpand redistribution` earlier used to error out when the table
would have been already expanded manually by the user or the table would have
been damaged. This commit changes this behaviour so that any expanded/broken
tables are skipped without affecting the redistribution for other tables.

New error message was added to gpexpand for search since we can fail with two
different messages in different cases.
execSingleton was changed to execSQLForSingleton in mgmt_utils.py to make it
work on 6x. Also, gpexpand.feature test was brought to a common standard by
adding 'rm -rf ~/gpAdminLogs/gpinitsystem*'. Without this line a failing
test would cause other tests to fail.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66185

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1137909

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1137910

**Comment by RekGRpth** [comment]
1) Format PR description according to guidelines.
2) If this is a cherry-pick, then where is the attribution?
3) Also, in the commit message you must indicate the cherry-pick of which commit it was.
4) Where is behavior test according steps to reproduce?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66436

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1149887

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1149888

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1149885

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1149881

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1149886

File: gpMgmt/bin/gpexpand [hunk 1]
```diff
@@ -10,6 +10,8 @@ import copy
 import datetime
 import os
 import random
+import re
+import re
```
Comment (RekGRpth) [inline gpMgmt/bin/gpexpand]: ?

Comment (dnskvlnk) [reply gpMgmt/bin/gpexpand]: Fixed

**Comment by RekGRpth** [comment]
I think the resulting patch should consist of just two cherry-picked commits.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66491

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1152325

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1152326

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1152324

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1152323

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1152319

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1152320

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66553

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1155495

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1155496

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66646

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161061

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161062

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161060

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161059

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161056

**Comment by RekGRpth** [comment]
> This commit is a cherry-pick of the community PR https://github.com/greenplum-db/gpdb/pull/15961.

1) The hash of the original commit must be specified, not a link to the pull request.
2) Why not two commits?

**Comment by swork12** [comment]
bender build

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66708

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1164085

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1164086

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1164084

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1164083

**Comment by RekGRpth** [comment]
> This commit is a cherry-pick of https://github.com/arenadata/gpdb/commit/b813c136ec18334f7d7098cc8d4564f1596d4c70.

That commit does not contain test and has different description.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66746

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1167685

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66750

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1167952

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1167944

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1167953

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66830

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1171285

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1171286

File: gpMgmt/bin/gpexpand [hunk 2]
```diff
@@ -2110,7 +2111,17 @@ class ExpandTable():
         # check is atomic in python
         if not cancel_flag:
             if sql:
-                dbconn.execSQL(table_conn, sql)
+                try:
+                    dbconn.execSQL(table_conn, sql)
+                except DatabaseError as e:
+                    if re.search('DETAIL:\s*table has already been expanded', str(e)):
+                        logger.info('Table %s.%s seems to be already expanded, marking as done' % (self.dbname, self.fq_name))
+                    elif re.search('\s*could not open relation with (OID \d+ )?', str(e)) or re.search('\s*relation not found (OID \d+ )?', str(e)):
```
Comment (RekGRpth) [inline gpMgmt/bin/gpexpand]: I think, the change
```go
 or re.search('\s*relation not found (OID \d+ )?', str(e))
```
should be described in commit message

File: gpMgmt/test/behave/mgmt_utils/gpexpand.feature [hunk 3]
```diff
@@ -660,3 +660,33 @@ Feature: expand the cluster by adding more segments
         And verify that the path "gpperfmon/logs" in each segment data directory does not exist
         And verify that the path "promote" in each segment data directory does not exist
         And verify that the path "db_analyze" in each segment data directory does not exist
+
+    @gpexpand_no_mirrors
+    @gpexpand_segment
+    Scenario: gpexpand should skip already expanded/broken tables when redistributing
+        Given the database is not running
+        # need to remove this log because otherwise SCAN_LOG may pick up a previous error/warning in the log
+        And the user runs command "rm -rf ~/gpAdminLogs/gpinitsystem*"
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/gpexpand.feature:669]: I think, these changes also should be described in commit message

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 4]
```diff
@@ -3418,6 +3419,23 @@ def impl(context):
...
             raise Exception("table %s has not finished expanding" % row[0])
 
+
+@then('table "{table_name}" {expansion_state} be marked as expanded')
+def impl(context, table_name, expansion_state):
+    dbname = 'postgres'
+    conn = dbconn.connect(dbconn.DbURL(dbname=dbname), unsetSearchPath=False)
+    try:
+        query = """select status from gpexpand.status_detail where fq_name='public.{0}'""".format(table_name)
+        result = dbconn.execSQLForSingleton(conn, query)
```
Comment (RekGRpth) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:3429]: I think, this change also should be described in commit message

File: gpMgmt/test/behave/mgmt_utils/gpexpand.feature (hunk 3 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/gpexpand.feature:669]: This line presents in all tests above, so it should have been there from the start, I don't think it has to be described

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 4 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:3429]: I don't think that such change should be described

Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:3429]: Done

File: gpMgmt/test/behave/mgmt_utils/gpexpand.feature (hunk 3 above)
Comment (dnskvlnk) [reply gpMgmt/test/behave/mgmt_utils/gpexpand.feature:669]: Done

File: gpMgmt/bin/gpexpand (hunk 2 above)
Comment (dnskvlnk) [reply gpMgmt/bin/gpexpand]: Done

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/66855

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173364

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173365

**Review by RekGRpth — DISMISSED** [review]

File: gpMgmt/bin/gpexpand (hunk 2 above)
Comment (andr-sokolov) [reply gpMgmt/bin/gpexpand]: @dnskvlnk , explain please why we need two conditions. The PR author believes that one condition is enough for 6X_STABLE https://github.com/greenplum-db/gpdb/pull/15892/files

Comment (dnskvlnk) [reply gpMgmt/bin/gpexpand]: This was a message from 7x, so I removed it

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67102

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1186726

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1186727

**Comment by RekGRpth** [comment]
cherry-pick only one commit with updated message, please

**Comment by dnskvlnk** [comment]
@RekGRpth the second commit will be squashed

**Comment by RekGRpth** [comment]
> @RekGRpth the second commit will be squashed

But wouldn't the authorship of the original commit be lost?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67151

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1189750

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1189751

**Review by RekGRpth — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67219

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193058

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193059

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1193048

**Pull request merged**